Data sub patch

austinpaine opened the pull request:
Fixed data leaking before user logs in.  Created subscription manager to handle subscribing users to each data source as they add them to their dashboard instead of sending all the data all of the time.

cawasser committed:
- added worldwind widget type (with static data embedded in the client)

cawasser committed:
- timeline!
- task-service (tasks and links between them)
- new vanilla_default! (for the "tasks-service" which generates timelines)
- some UI tweaks (e.g., WW and timeline widgets fill the width)

cawasser committed:
- thumbnails for WW and timeline widgets
- worldwind-react-globe.js

cawasser committed:
- fixed thumbnail name for timeline-widget

Commit:
started carousel add

cawasser committed:
- make worldwind-widget compatible with :data-format/cont-n (world-wide map)

cawasser committed:
- turn off some extraneous tracing

cawasser committed:
- split widget adding modal into two: 1) add by data-source first, 2) add by widget first

cawasser committed:
- added some "types" to the tasks, which translate to "colors" in the widget

cawasser committed:
- added a "hand-drawn" diagram display widget using https://rough-cljs.netlify.com. will need lots of work to be sufficiently flexible for dynamic use (scaling, translation, etc.)

cawasser committed:
- changes continent-service to provide just the city locations
- modified continent-map to add in the series information needed to draw a 2d map

Commit:
Merge branch 'ww-timeline-carousel' of https://github.com/cawasser/vanilla into ww-timeline-carousel

cawasser committed:
- changes australia-map-service to provide just the city locations
- modified australia-map to add in the series information needed to draw a 2d map

cawasser committed:
- added terminal_location_service.clj to provide terminal lat/lon form an excel spreadsheet

cawasser committed:
- added terminal-location-service and signal-path-service to database

cawasser committed:
- added signal_path_service.clj

cawasser committed:
- added signal_path_service.clj

cawasser committed:
- added Bandwidth as the "weight" factor on the signal-path
- reduced the terminal count to 100 because continent-map wasn't drawing when it was the full dataset (approx. 1200)

cawasser committed:
- dummy data for terminal locations and signal paths

cawasser committed:
- moved the excel filename into the supporting namespace to simplify changes

Commit:
carousel mostly working, hardcoded widget, no adding

cawasser committed:
- updated to add beams sheet

cawasser committed:
- remove some tracing

cawasser committed:
- added beam-location-service to server

cawasser committed:
- remove tracing

cawasser committed:
- add ^{:key} metadata to remove a React warning

cawasser committed:
- moved layer setup to vanilla.mapping.layer-management

cawasser committed:
- remove links from the task-service data (un-needed)

cawasser committed:
- added a Missions tab to Demo.xlsx
- task-service reading data from Demo.xlsx
- timeline no longer needs to parse data as an "#inst" is returned from the server

Commit:
pulled and merged with chris changes

Commit:
comments

Commit:
carousel working, hardcoded sources/charts

cawasser committed:
- pulling beam-layer data from the beam-location-service

cawasser committed:
Merge branch 'ww-timeline-carousel' of https://github.com/cawasser/vanilla into ww-timeline-carousel

cawasser committed:
- tidy up carousel

cawasser committed:
- move buttons before the Slider in the hopes of better fitting the widget size (still not there)

cawasser committed:
- changes to the structure of several worksheet in the Demo.xlsx file (signal-path still incorrect)
- show beam name on the layer
- sort task and signal-path data at the server

cawasser committed:
- demo.xlsx/SCN_NETWORK_CARRIER_VW updated to the new format

cawasser committed:
- css adjustments for widget sizing

cawasser committed:
- white-dot.png for WorldWind

cawasser committed:
- added labels for gantt bars

cawasser committed:
- fixed bug from bad refactor into widget-title-bar

cawasser committed:
- removed ":minWidth" form arearange-chart
- updated beam locations and radii in Demo.xlsx

cawasser committed:
- added a BlackHammer background image to the page

cawasser committed:
Merge branch 'ww-timeline-carousel' of https://github.com/cawasser/vanilla into ww-timeline-carousel

cawasser committed:
- switched to #splash selected since WorldWind uses a canvas too

Commit:
route added

Commit:
Merge branch 'ww-timeline-carousel' into data-sub-patch

Commit:
datasource post tweaks

Commit:
not working but closer, runs

cawasser committed:
- added more cities. we should look into getting these from an open Internet source.

cawasser committed:
- some documentation updates

cawasser committed:
- switch to load-workbook-from-resource so this will work from an uberjar (in another folder) or a docker container

cawasser committed:
- updated to jquery@^3.5.0 (lower-level dependency) due to moderate severity vulnerability flagged by github (published 2020-40-29, https://github.com/cawasser/vanilla/network/alert/package-lock.json/jquery/open)

cawasser committed:
- rearranged the splash image on index.html to work better as a background

Commit:
Merge branch 'ww-timeline-carousel' into data-sub-patch

Commit:
changed cljs data-source socket init connect to after login to fix leak

Commit:
changed services get to after login

Commit:
nearly working, need data push on new-widget add, and cleanup on delete

Commit:
nearly nearly working. figuring out websocket channels

Commit:
data-source subscriptions by user are working

Commit:
merge with master

Commit:
cleanup

Commit:
cleanup

austinpaine commented on src/clj/dashboard_clj/components/websocket.clj:
```diff
@@ -67,3 +81,129 @@
                          :handler           client-ev-handler
                          :webserver-adapter webserver-adapter
                          :options           options}))
+
```
Ignore huge comment below.  left it in for future reference if we need to see how some of these messages/events are structured

austinpaine commented on src/cljs/dashboard_clj/core.cljs:
```diff
@@ -14,9 +14,11 @@
-    (asyncm/go-loop []
+    (async/go-loop []
       (let [{:keys [event id ?data send-fn]} (async/<! ch-recv)]
 
```
noticed this typo that was in master (asyncm vs async), surprised the compiler didnt catch this, and it seemed to function as normal?  Changed to async, but no noticeable effects either way.  Strange

austinpaine commented on src/cljs/vanilla/core.cljs:
```diff
@@ -79,21 +78,6 @@
 
 
 
```
This has been pulled out of core and moved to update-layout

austinpaine commented on src/cljs/vanilla/update_layout.cljs:
```diff
@@ -64,8 +86,9 @@
+                 ;(prn ":set-layout CONVERTED:  " converted-data)
+
+                 (data-source-subscribe (mapv #(:data-source %) converted-data))
 
```
the new data-source-subscribing happens every time the layout is updated

austinpaine commented on src/cljs/vanilla/widgets/pie_chart.cljs:
```diff
@@ -46,10 +46,10 @@
+    ;  " //// (ret) " ret)
 
     ret))
 
```
noticed this print that got missed at some point and commented it out for cleanliness

cawasser reviewed (changes requested):
Need to add some processing to basic-widget/[:remove-widget]-handler which should remove the "unused" data-source(s) from re-frame just to keep things tidy. I noticed that deleting a widget does NOT remove the data-source. Assuming the server would keep sending updates too, but didn't check for sure.

Also, since it is possible to subscribe to the same source from multiple widgets, we probably need to keep track of how many times a given source is subscribed and only remove it when the LAST widget is removed (count drops to 0) AND tell the server the client is no longer interested so it stops sending.

austinpaine commented:
Oooh good catches, will implement

Commit:
removed data-source after widget removal

cawasser commented on src/cljs/vanilla/widgets/basic_widget.cljs:
```diff
@@ -23,7 +28,8 @@
     (delete-widget widget-id)
-    (assoc db :widgets (remove #(= (:key %) widget-id) (:widgets db)))))
+    (assoc db :widgets (remove #(= (:key %) widget-id) (:widgets db))
+              :data-sources (dissoc (:data-sources db) (get-data-source @(rf/subscribe [:active-widgets-by-user]) widget-id)))))
```
Interesting. Even without reference counting, a "removed" subscription will be resubscribed if another widget is still subscribing. It looks like this happens a part of layout management. We'll want to keep an eye on if this causes any thrashing.

austinpaine replied:
You're quick!  Yep was just going to explain that it still worked based on how widgets get sent over if there are multiple widgets with the same data source.  Besides a flash as the new data updates get sent, no noticeable issues.  Might be worth a check once we turn scheduling back on eventually.

gpierce3 reviewed (changes requested):
Noticed two things:

1. Looking at app-db when running in dev mode.  Picked two widgets.  Logged out.  data sources still showed in :datasources.  Is that what we want?

2. I picked a spectrum trace widget and then a carousel.  For kicks, spun thru the carousel.  The second widget was blank except for the chart title 'Sat2 Power'.  I then deleted the spectrum trace.  Now none of the chart info shows in carousel, just chart name.  It still 'knows' there are three widgets in the carousel. This happened in dev/jar/docker.  Went back to master branch and retried in dev mode.  Worked okay.

cawasser committed:
- pulled data-source-subscribe into it's own ns for use throughout the client
- added call to subscribe to hard-coded data-sources for layer-management and carousel.cljs (does NOT unsubscribe - fix in next version)
- no feature regression from prior version

cawasser reviewed (commented):
I added a fix for the (hard-coded) subscriptions for Carousel and WorldWind. Otherwise we get a feature regression (things stop working). This should also address gpierce3 comment #2 above.

Please double-check my work here.

Commit:
logout subscription cleanup

austinpaine commented:
Oh good catch with the logout data-sources George.  I had cleanup setup for when the user closed the tab (websocket connection close), but not logout.  Went down too hard of a path for a bit fixing this, but then realized i can just dissoc the data-sources and services easily in the logout event.  Done!

As for carousel, thats because the "data-source" for carousel is entirely fake and mocked up right now.  I hadnt tested this functionality with that widget because i didnt expect it to work at all.  I just took on the automatic carousel widget task today which will cover getting these subscriptions sorted out for carousel as well.

Commit:
Merge branch 'data-sub-patch' of https://github.com/cawasser/vanilla into data-sub-patch

austinpaine commented:
Thanks Chris for sticking in that quick fix!  Just looked it over and tested it and seems to work out great.  Will deal with getting it sorted out proper with the current carousel task

cawasser commented:
We'll need to make some DB updates to add support for the "hard-coded" subscriptions for WW and Carousel, they aren't currently exposed that way. This will set things up for the "add a source" feature coming soon.

chadarm reviewed (approved):
Seems like a lot of good moves made in this PR, makes everything feel more organized. 
Great work with commenting and explaining functions, super easy to follow what's happening.

cawasser commented on src/clj/dashboard_clj/components/websocket.clj:
```diff
@@ -67,3 +81,129 @@
                          :options           options}))
+
+
+(comment
```
should probably add a "real" comment describing what/why this is here

austinpaine replied:
added

Commit:
comments

Commit:
merge with live demo master

austinpaine commented:
Merged with latest master branch after live demo merge.  Fixed the only two conflicts.  Ready to merge.

cawasser commented on src/cljs/vanilla/mapping/layer_management.cljs:
```diff
@@ -104,7 +104,8 @@
   ; TODO: this is a hack for the following hack (does NOT unsubscribe to sources when widget closes)
   (ds/data-source-subscribe [:beam-location-service :terminal-location-service])
```
need to update the subscriptions to include :ka-beam-location-service

cawasser commented on src/cljs/vanilla/widgets/carousel.cljs:
```diff
@@ -295,10 +295,10 @@
   ; TODO: this is a hack for the following hack (does NOT unsubscribe to sources when widget closes)
   (ds/data-source-subscribe [:spectrum-traces :power-data])
```
need to update the subscriptions to be [:channel-power-1000-service :channel-power-2000-service :channel-power-3000-service :channel-power-4000-service]

Commit:
fixed subscriptions
